OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs

```diff
@@ -11,55 +11,146 @@
-		public readonly string Format = "{0}: {1}";
-		public readonly TimerOrder Order = TimerOrder.Descending;
+		[Translate] public readonly string NoTeamText = "No Team";
+		[Translate] public readonly string TeamText = "Team {0}";
 
-		readonly IEnumerable<SupportPowerInstance> powers;
-		Pair<string, Color>[] texts;
+		readonly Dictionary<Player, SupportPowerManager> spManagers;
+		readonly SortedDictionary<int, List<Player>> playersByTeam;
+
+		readonly string fontName = "Bold";
+		readonly SpriteFont font;
+
+		readonly float yIncrement;
+		readonly float padding;
+
+		struct Text
+		{
+			public Color Color;
+			public string Team;		// Used only for team label
+			public string Player;
+			public string Power;
+			public string RemainingTime;
+		}
+
+		List<Text> display;
+		float maxPlayerWidth;
+		float maxPowerWidth;
+
+		World world;
+		int lastTick;
 
 		[ObjectCreator.UseCtor]
 		public SupportPowerTimerWidget(World world)
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
-				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
-				.SelectMany(s => s.Trait.Powers.Values)
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			spManagers = new Dictionary<Player, SupportPowerManager>();
+			foreach (var tp in world.ActorsWithTrait<SupportPowerManager>().Where(tp => !tp.Actor.Owner.NonCombatant))
+				spManagers[tp.Actor.Owner] = tp.Trait;
+
+			playersByTeam = new SortedDictionary<int, List<Player>>();
+			foreach (var team in spManagers.Keys.GroupBy(p => world.LobbyInfo.ClientWithIndex(p.ClientIndex).Team))
+			{
+				playersByTeam[team.Key] = new List<Player>();
+				playersByTeam[team.Key].AddRange(team);
+			}
+
+			font = Game.Renderer.Fonts[fontName];
+			yIncrement = font.Measure(" ").Y + 5;
+			padding = font.Measure("  ").X;
+
+			display = new List<Text>();
+			this.world = world;
 		}
 
 		public override void Tick()
 		{
-			texts = powers.Select(p =>
+			if (lastTick == world.WorldTick)
+				return;
+
+			lastTick = world.WorldTick;
+			display = new List<Text>();
+			maxPlayerWidth = 0;
+			maxPowerWidth = 0;
+			foreach (var team in playersByTeam)
 			{
-				var time = WidgetUtils.FormatTime(p.RemainingTime, false);
-				var text = Format.F(p.Info.Description, time);
-				var color = !p.Ready || Game.LocalTick % 50 < 25 ? p.Instances[0].Self.Owner.Color.RGB : Color.White;
-				return Pair.New(text, color);
-			}).ToArray();
+				var teamDisplay = new List<Text>();
+				foreach (var player in team.Value)
+				{
+					var spiList = spManagers[player].Powers.Values
+						.Where(i => i.Instances.Any() && i.Info.DisplayTimer && !i.Disabled);
+
+					if (!spiList.Any())
+						continue;
+
+					var playerWidth = font.Measure(player.PlayerName).X;
+					if (playerWidth > maxPlayerWidth)
+						maxPlayerWidth = playerWidth;
+
+					var showPlayer = true;
+					foreach (var spi in spiList)
+					{
+						var powerWidth = font.Measure(spi.Info.Description).X;
```

Comment:
Likewise here - you know all the possible support power names before you start, so you can build a dictionary at startup instead of measuring everything every tick.